client/components/tinymce/plugins/media/plugin.jsx

```diff
@@ -399,17 +398,33 @@ function mediaButton( editor ) {
 		delete media.caption;
 		media = assign( {}, parsed.media, media );
 
+		// Compute a ratio used to order media sizes
+		const computeRatio = ( { width = media.width, height = media.height } ) => {
+			return Math.min(
+				( width / media.width ) || Infinity,
+				( height / media.height ) || Infinity
+			);
+		};
+
 		// Determine the next usable size
-		let sizeIndex = SIZE_ORDER.indexOf( parsed.appearance.size );
+		const sizeRatios = SIZE_ORDER
+			.map( size => MediaUtils.getThumbnailSizeDimensions( size, site ) )
+			.map( computeRatio );
```

Comment:
We should avoid iterating twice, here and below, by combining into a single map:

```js
const sizeRatios = SIZE_ORDER
	.map( size => computeRatio( MediaUtils.getThumbnailSizeDimensions( size, site ) ) );
```

And filter:

```js
const possibleSizes = SIZE_ORDER
	.filter( ( size, index ) => {
		return sizeRatios[ index ] <= 1 && (
			increment > 0
				? sizeRatios[ index ] > displayedRatio
				: sizeRatios[ index ] < displayedRatio
		);
	} );
```